Adds --forceupdate to teosd

sr-gi opened the pull request:
If a tower hasn't been running for a long time and the backend runs in pruned mode it could be the case that by the time the tower comes back online, the Last known block by the tower is not being known by the node anymore. In this situation, the tower cannot bootstrap normally, given the cache cannot be populated.

This commits adds a new argument to teosd (`--forceupdate`) that can be used to force a tower to update its last known block to the earliest known block by the backend under this situation. Notice that doing so may make the tower miss some of its state transitions (the ones triggered by missed blocks), so this must be done as a last resource.

Fixes #209 
Supersedes #212 

## Testing

This can be tested easily in regtest by running `bitcoind` with the (hidden, undocumented) `-fastprune` modifier. That would make blocks smaller and easier to prune. After that, you can prune the chain using `bitcoin-cli pruneblockchain x` where `x` is your target prune height. The pruning is not likely to happen to the exact block you specified, given block files are pruned as a whole, so the chain would be pruned to the closest value that allows a bock file to be deleted.

So the flow of testing should be:

1. Start `teosd` and mine a few blocks so there is a `last_known_block` known by the tower
2. Shut down the tower
3. Mine a few thousand more blocks
4. Call `bitcoin-cli pruneblockchain x` and checked the returned `y` value
5. If `y - IRREVOCABLY_RESOLVED < last_known_block_height`, continue, otherwise go back to `3.`
6. Run `teosd` again and test it with and without `--forceupdate`

sr-gi commented:
I've tested this manually and it seems to run just fine.

In order to test this you'd need a tower that has a last known block at height `n` and a node with a chain that has pruned that same block. I haven't been able to test this on regtest given pruning is based on how much block data we are holding (not how many blocks) and regtest blocks are mainly empty, so the threshold is never met. However, I've been able to test this using mainnet and an old tower state.

I may add some E2E testing for this if I manage to create a test where full blocks are created, so eviction can be triggered.

mariocynicys commented on teos/src/main.rs:
```diff
@@ -193,7 +190,7 @@ async fn main() {
             .validate(block_hash)
             .unwrap()
```
Ummm, doesn't this fail as well when we are out of blocks?

sr-gi replied:
Actually, it does not because the headers chain is preserved even if the blocks are pruned.

mariocynicys commented on teos/src/main.rs:
```diff
@@ -218,7 +215,51 @@ async fn main() {
+                    .unwrap();
+                if let Ok(blocks) = get_last_n_blocks(
+                    &mut poller,
+                    bitcoin_cli
+                        .deref()
+                        .get_header(&hash, Some(height as u32))
+                        .await
+                        .unwrap()
+                        .validate(hash)
+                        .unwrap(),
```
nit: can we move this into a variable and call it `tip`.

mariocynicys commented on teos/src/main.rs:
```diff
@@ -218,7 +215,51 @@ async fn main() {
+        .await
+    {
+        Ok(blocks) => blocks,
+        Err(_) => {
```
If we go into this branch we would return some last_n_blocks whose tip isn't the same as the `tip` variable we are using in the initialization of the tower components (watcher, etc...).

mariocynicys commented on teos/src/main.rs:
```diff
@@ -218,7 +215,51 @@ async fn main() {
+                let hash = rpc
+                    .get_block_hash(height + IRREVOCABLY_RESOLVED as u64)
+                    .unwrap();
+                if let Ok(blocks) = get_last_n_blocks(
```
These `n` blocks won't be delivered to the tower components, right?

mariocynicys reviewed (commented):
I'm confused about what the `chain_monitor` would do given it's being passed an old/pruned tip. Is `poll_best_tip()` gonna just skip the pruned blocks between `tip` and the best tip?

sr-gi commented:
> I'm confused about what the `chain_monitor` would do given it's being passed an old/pruned tip. Is `poll_best_tip()` gonna just skip the pruned blocks between `tip` and the best tip?

Yeah, you're right. I've come up with a better solution than this, will force push asap

sr-gi commented:
This should make more sense now

mariocynicys commented on teos/src/main.rs:
```diff
@@ -186,14 +183,50 @@ async fn main() {
+        // If we are running in pruned mode some data may be missing (if we happen to have been offline for a wile)
+        if let Some(prune_height) = rpc.get_blockchain_info().unwrap().prune_height {
+            if last_known_header.height < prune_height as u32 {
+                log::debug!(
```
I think this should be logged as a warning.

mariocynicys commented on teos/src/main.rs:
```diff
@@ -186,14 +183,50 @@ async fn main() {
+
+        // If we are running in pruned mode some data may be missing (if we happen to have been offline for a wile)
+        if let Some(prune_height) = rpc.get_blockchain_info().unwrap().prune_height {
+            if last_known_header.height < prune_height as u32 {
```
We would still fail if some of the 100 blocks that proceeds the last known one are pruned.

The check should be: `last_known_header.height - 100 < prune_height as u32`

sr-gi replied:
`<=`, but you're right

sr-gi replied:
Oh, NVM, I was thinking about the positive case (`>=`)

sr-gi commented:
I fixed suggestions and updated some of the logs since I didn't like them much.

I've also updated the PR description with guidelines for how to test this in regtest.

mariocynicys reviewed (commented):
Some nits, looks good otherwise.
Didn't manually test it yet though.

sr-gi commented:
Nits should be fixed. I'd be nice to get a tACK

mariocynicys commented:
> Nits should be fixed. I'd be nice to get a tACK

Yeah will work on that today.

mariocynicys commented on teos/src/main.rs:
```diff
@@ -186,14 +183,53 @@ async fn main() {
+        if let Some(prune_height) = rpc.get_blockchain_info().unwrap().prune_height {
+            if last_known_header.height - IRREVOCABLY_RESOLVED < prune_height as u32 {
+                log::warn!(
+                    "Cannot load blocks in the range {}-{}. Chain has gone to far out of sync",
```
typo: "~to~too far our of sync"

mariocynicys commented on teos/src/main.rs:
```diff
@@ -186,14 +183,53 @@ async fn main() {
+
+        // If we are running in pruned mode some data may be missing (if we happen to have been offline for a while)
+        if let Some(prune_height) = rpc.get_blockchain_info().unwrap().prune_height {
+            if last_known_header.height - IRREVOCABLY_RESOLVED < prune_height as u32 {
```
This has an off-by-one error :joy:, should be `last_known_header.height - IRREVOCABLY_RESOLVED + 1 < prune_height as u32`.

Let's say we last witnessed block 3000, then we will want to load blocks 2901-3000 (100 blocks). `prune_height` indicates the lowest block ***we have*** stored, so if `prune_height` is 2901 we should be fine, but `last_known_header.height - IRREVOCABLY_RESOLVED` will be 2900, which would make the check fail.

sr-gi replied:
You're right

mariocynicys commented on teos/src/main.rs:
```diff
@@ -186,14 +183,53 @@ async fn main() {
+            if last_known_header.height - IRREVOCABLY_RESOLVED < prune_height as u32 {
+                log::warn!(
+                    "Cannot load blocks in the range {}-{}. Chain has gone to far out of sync",
+                    last_known_header.height - IRREVOCABLY_RESOLVED,
```
This should be `last_known_header.height - IRREVOCABLY_RESOLVED + 1`.

mariocynicys reviewed (changes requested):
tACK

Some comments (again :p) inline, looks good otherwise.

sr-gi commented:
Addressed comments

mariocynicys reviewed (approved):
tACK [c7f9f0a](https://github.com/talaia-labs/rust-teos/pull/216/commits/c7f9f0a61211d72df7101fecf0715d22a088e19b)
